### PR TITLE
#162949081 Users should be able to follow each other

### PR DESCRIPTION
**Pull request by Paulvitalis200**
### What does this PR do?
Adds the ability for a user to favorite and unfavorite articles.
A user will also be able to view a list of all his/her favorite articles

### Description of Task to be completed?
- [x] Create Follower model
- [x] Add tests
- [x] Create model for Favorite relation
- [x] Add Followers serializer
- [x] Create Followers views
- [x] Have the following endpoints working:
     - POST `api/v1/profiles/<username>/follow/`
     - GET `api/v1/following/`
     - DELETE `api/v1/follow/<username>`
- [x] Add following count
- [x] Display following count on the profile of a user

### How should this be manually tested?
1. Clone this repository on your local machine using 

     `git clone https://github.com/andela/ah-robotics.git`
2. Change your working directory to the project's directory using 
     
     `cd ah-robotics`
3. Change to the specific branch `ft-favorite-followers-162949081` using 

     `git checkout ft-followers-162949081`
4. Create a virtual environment using the following command: 
     
     `python3 -m virtualenv env`
5. Activate the virtual environment using 
     
     `source env/bin/activate`
6. Install all the requirements using 
     
     `pip install -r requirements.txt`

    #### Database setup

    Ensure [PostgreSQL](https://www.postgresql.org/) is installed and is running, follow these steps to create a new user and a database which will be used in the application.

    1. Start the postgres database server using the command:
      
        `pg_ctl -D /usr/local/var/postgres start`
    
    2. Run psql interactive terminal
    
        `psql postgres`
    
    3. Create a new user:
    
        `CREATE USER sample_username WITH PASSWORD 'sample_password';`
    
    4. Grant privileges to the user:
    
        `ALTER USER sample_username CREATEDB;`
    
    5. Create a database:
    
        `CREATE DATABASE sample_database_name WITH OWNER sample_username;`
    
    6. Exit psql interactive with the following command: `\q`
    
7. Create a **.env** file and add the following configuration variables

    ```source-json
          source env/bin/activate
          export DB_NAME="database_name"
          export DB_USER="sample_username"
          export DB_PASS="sample_password"
          export DB_HOST="host_name_eg_localhost"
          export DB_PORT="port_number_eg_5432"
          export TIME_DELTA="30"
          export EMAIL_HOST="smtp.gmail.com"
          export EMAIL_PORT=587
          export EMAIL_HOST_USER="user@example.com"
          export EMAIL_HOST_PASSWORD="email_password"
          export PASSWORD_RESET_URL="api/v1/reset_password"
          export CLIENT_DOMAIN="127.0.0.1:8000"
          export CLOUDINARY_NAME="your_cloudinary_name"
          export CLOUDINARY_KEY="your_cloudinary_key"
          export CLOUDINARY_SECRET="your_secret_key"
    ```
    
    > **Note**: The values of some of these variables are private namely the EMAIL_HOST_USER, EMAIL_HOST_PASSWORD and the CLOUDINARY VALUES.  Use your GMAIL email address and password for the  EMAIL_HOST_USER and EMAIL_HOST_PASSWORD respectively. To get cloudinary values, create an account with cloudinary and obtain the values.

8. Source .env to set the configuration variables in the terminal:

      `source .env`
9. Run migrations through the terminal to create the tables in the database:
      
      (i) Make the migrations using: 
      
      `python manage.py makemigrations`
      
      (ii) Apply the migrations using: 
      
      `python manage.py migrate`

10. Run the following command to collect static files.

    `python manage.py collectstatic`
      
#### Testing the features

1. Run the server from the terminal using: 
    
    `python manage.py runserver`
2. Ensure [Postman](https://www.getpostman.com/) is installed, open it on your local machine and Paste the URL `http://127.0.0.1:8000/api/v1/users/` to first sign up a user.
3. On the **Headers** section add **Content-type** as the KEY and **_application/json_** as the VALUE.
4. Head over to the **Body** section and select the **raw** option and the add the following payload to the body:

    ```
        {
            "user": {
                "username": "sample_username",
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```
5. Once all the fields are inserted, ensure that it is a **POST** request and click the **Send** button.
6. Sign up another user using the same payload format in step 4 but with different fields to have at least two users in the system.
7. Log in the first user using the following endpoint `http://127.0.0.1:8000/api/v1/users/login/` and insert the following payload to the body section.
    
    ```
        {
            "user": {
                "email": "sample_email",
                "password": "sample_password"
                }
        }
    ```
8. After logging in the screen should look as follows:
    
    image

9. Copy the token and click on the **Headers** section. In the **Key** column, type in the word **Authorization**. And in its corresponding value in the **Value** column, write the word __token__, add a space and paste in the token that you copied from the login response. The **Headers** section should look as follows after doing so:
    
    image

10. To follow the second user, use the URL: http://127.0.0.1:8000/api/v1/profiles/username/follow. For example, to follow a user with the username of paulotieno, use http://127.0.0.1:8000/api/v1/profiles/paulotieno/follow
Make sure the body section is empty by removing any payload which is within it as you do not require any payload.
Ensure the method is a **POST** and hit **Send**. After sending the request, Change the URL to that of the profiles, which is http://127.0.0.1:8000/api/v1/profiles/ and do a **GET** request and send. You should see the **following** field of the specific user as **true** and that user's **followers** value should increment by 1.

11. To unfollow the second user, use the URL http://127.0.0.1:8000/api/v1/profiles/username/unfollow. For example:
http://127.0.0.1:8000/api/v1/profiles/paulotieno/unfollow. 
Make sure the body section is empty by removing any payload which is within it as you do not require any payload.
Change the method to **DELETE** and hit **Send**. After sending the request, change the URL to that of the profiles which is http://127.0.0.1:8000/api/v1/profiles/ and do a **GET** request and send. You should see the **following** field of the specific user as **false** and that user's **followers** value should decrease by 1.

12. To view the list of users a specific user is following, use the URL http://127.0.0.1:8000/api/v1/profiles/username/following/. For example:
http://127.0.0.1:8000/api/v1/profiles/teatoller/following/.

Change the method to a **GET** and hit send.

13. To view a specific users followers, use the URL http://127.0.0.1:8000/api/v1/profiles/username/followers/. For example:
http://127.0.0.1:8000/api/v1/profiles/teatoller/followers/.
Change the method to a **GET** and hit send.

#### Running tests
1. Head over to the terminal and ensure that the server is not running and you are in the project root folder.
2. Run the tests by typing in the following command: 
    
    `python manage.py test`

### Any background context you want to provide?
None

### What are the relevant pivotal tracker stories?
[#162949081](https://www.pivotaltracker.com/n/projects/2235171/stories/162949081)

### Screenshots

> **Following a user**
image

> **Following a user that is already followed**
image

> **Trying to follow a non-existent user**
image

> **Unfollowing a user**
image

> **Unfollowing a user that is not within the followers**
image

> **Getting a list of the users a specific user is following**
image

> ** Getting the list of followers of a specific user**
image

> **Viewing the following status and followers of other users**
image

## Questions:
Questions and comments are welcome.

### CODE REVIEW

File: authors/apps/followers/models.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+from django.db import models
+from authors.settings import AUTH_USER_MODEL
```
Comment (sharkdevs) [inline authors/apps/followers/models.py:3]: separate your imports

File: authors/apps/followers/models.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+from authors.apps.authentication.models import User
+
+# Create your models here.
```
Comment (sharkdevs) [inline authors/apps/followers/models.py]: You can remove this comment because that is definitely a model

**Review by sharkdevs — COMMENTED** [review]
Good work, it is working fine on my side. Just consider the changes highlighted below.

**Comment by jomasim** [comment]
@Paulstar200  Good work I have reviewed your work and managed to test it locally.

File: authors/apps/followers/models.py (hunk 1 above)
Comment (Paulvitalis200) [reply authors/apps/followers/models.py:3]: Thank you. I have fixed that

File: authors/apps/followers/models.py (hunk 2 above)
Comment (Paulvitalis200) [reply authors/apps/followers/models.py]: I have fixed this

**Comment by Paulvitalis200** [comment]
> @Paulstar200 Good work I have reviewed your work and managed to test it locally.

Thank you for the feedback.

**Review by waracci — COMMENTED** [review]
Works as expected! Good one.

**Review by johnwayodi — APPROVED** [review]
Nice work. Everything works well.

**Comment by Teatoller** [comment]
Hello @Paulstar200 

On instruction ```11```

Add profile to ```http://127.0.0.1:8000/api/v1/username/unfollow``` 
  &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;&nbsp;&downarrow;
```http://127.0.0.1:8000/api/v1/profile/username/unfollow```

Without which the desired output fails.

```To unfollow the second user, use the URL http://127.0.0.1:8000/api/v1/username/unfollow. For example:
http://127.0.0.1:8000/api/v1/profiles/paulotieno/unfollow.
Change the method to DELETE and hit Send. After sending the request, 
change the URL to that of the articles which is 
http://127.0.0.1:8000/api/v1/profiles/ and do a GET request and send. 
You should see the following field of the specific user as false.
```
Also include in your instructions for user to clear the body textfields when a post does not require or call for a payload inorder to avoid the no response message.
Regards
Steven Ennis

**Comment by Paulvitalis200** [comment]
> Hello @Paulstar200
> 
> On instruction `11`
> 
> Add profile to `http://127.0.0.1:8000/api/v1/username/unfollow`
>                         ↓
> `http://127.0.0.1:8000/api/v1/profile/username/unfollow`
> 
> Without which the desired output fails.
> 
> ```
> http://127.0.0.1:8000/api/v1/profiles/paulotieno/unfollow.
> Change the method to DELETE and hit Send. After sending the request, 
> change the URL to that of the articles which is 
> http://127.0.0.1:8000/api/v1/profiles/ and do a GET request and send. 
> You should see the following field of the specific user as false.
> ```
> Also include in your instructions for user to clear the body textfields when a post does not require or call for a payload inorder to avoid the no response message.
> Regards
> Steven Ennis

Thank you for that. I have fixed it

**Pull request merged**